### PR TITLE
Make `chx.backward` not cause error even if backprop is not required

### LINKED ISSUE #7272: Make `chx.backward` not cause error even if backprop is not required
Context: https://github.com/chainer/chainer/issues/6612#issuecomment-495080834

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #7272

### CODE REVIEW

File: chainerx_cc/chainerx/backward.cc [hunk 1]
```diff
@@ -211,27 +213,34 @@ class BackwardImpl {
...
+        // destruction order problem.
+        static const std::shared_ptr<ArrayNode> kNullArrayNode{nullptr};
+
         for (const Array& output : outputs) {
-            if (!output.IsBackpropRequired(backprop_id)) {
-                throw ChainerxError{"Cannot start backprop from an array whose gradient is not required (on graph '", backprop_id, "')"};
+            if (output.IsBackpropRequired(backprop_id)) {
+                output_array_nodes_.emplace_back(internal::GetArrayBody(output)->GetArrayNode(backprop_id));
+            } else {
+                output_array_nodes_.emplace_back(kNullArrayNode);
```
Comment (hvy) [inline chainerx_cc/chainerx/backward.cc]: You can just write as follows (and avoid the static shared_ptr)?
```cpp
output_array_nodes_.reserve(outputs.size());
for (const Array& output : outputs) {
    output_array_nodes_.emplace_back(internal::GetArrayBody(output)->GetArrayNode(backprop_id));
}
```

File: chainerx_cc/chainerx/backward.cc [hunk 2]
```diff
@@ -211,27 +213,34 @@ class BackwardImpl {
...
             }
-            output_array_nodes_.emplace_back(internal::GetArrayBody(output)->GetArrayNode(backprop_id));
         }
 
-        // Check if backward is possible for the given graph, in this context.
-        // It is not possible if a graph from an outer scope has already been backpropped.
-        backprop_id.context().CheckBackpropAllowed(backprop_id);
+        if (!output_array_nodes_.empty()) {
```
Comment (hvy) [inline chainerx_cc/chainerx/backward.cc]: A minor but `output_array_nodes_` should have the same size here as `outputs`. If `outputs` is empty, this ctor body does essentially nothing. How about checking that in the beginning with an early return?

File: chainerx_cc/chainerx/backward_test.cc [hunk 3]
```diff
@@ -334,15 +334,29 @@ TEST_F(BackpropTest, BackwardWithComplicatedRanks) {
...
     auto xs = MakeFullArrays({1}, {2.0f, 3.0f});
     auto y1 = xs[0] * xs[1];  // without graph
-    EXPECT_THROW(Backward(y1), ChainerxError);
+    Backward(y1);
+
+    EXPECT_THROW(xs[0].GetGrad(), ChainerxError);
+    EXPECT_THROW(xs[1].GetGrad(), ChainerxError);
+}
+
+TEST_F(BackpropTest, BackwardFromArrayWithAndWithoutNode) {
```
Comment (hvy) [inline chainerx_cc/chainerx/backward_test.cc]: Are the comments in this test reversed ("with"/"without")?

Comment (niboshi) [reply chainerx_cc/chainerx/backward_test.cc]: Oops, I thought the comments were wrong, but I was wrong.

File: chainerx_cc/chainerx/backward.cc (hunk 1 above)
Comment (niboshi) [reply chainerx_cc/chainerx/backward.cc]: Fixed as suggested.

File: chainerx_cc/chainerx/backward.cc (hunk 2 above)
Comment (niboshi) [reply chainerx_cc/chainerx/backward.cc]: Done.

**Comment by niboshi** [comment]
PTAL.

**Comment by hvy** [comment]
Jenkins, test this please.

**Comment by hvy** [comment]
LGTM!

**Pull request merged**